### PR TITLE
feat: whatshap haplotag

### PR DESCRIPTION
**Pull request by pdimens**
### Description
Add a wrapper for `whatshap haplotag`, a submodule of `whatshap` that phases BAM files using a phased VCF file. It also initiates the framework for adding other `whatshap` modules.

### QC

* [x] I confirm that:

For all wrappers added by this PR, 

* there is a test case which covers any introduced changes,
* `input:` and `output:` file paths in the resulting rule can be changed arbitrarily,
* either the wrapper can only use a single core, or the example rule contains a `threads: x` statement with `x` being a reasonable default,
* rule names in the test case are in [snake_case](https://en.wikipedia.org/wiki/Snake_case) and somehow tell what the rule is about or match the tools purpose or name (e.g., `map_reads` for a step that maps reads),
* all `environment.yaml` specifications follow [the respective best practices](https://stackoverflow.com/a/64594513/2352071),
* wherever possible, command line arguments are inferred and set automatically (e.g. based on file extensions in `input:` or `output:`),
* all fields of the example rules in the `Snakefile`s and their entries are explained via comments (`input:`/`output:`/`params:` etc.),
* `stderr` and/or `stdout` are logged correctly (`log:`), depending on the wrapped tool,
* temporary files are either written to a unique hidden folder in the working directory, or (better) stored where the Python function `tempfile.gettempdir()` points to (see [here](https://docs.python.org/3/library/tempfile.html#tempfile.gettempdir); this also means that using any Python `tempfile` default behavior works),
* the `meta.yaml` contains a link to the documentation of the respective tool or command,
* `Snakefile`s pass the linting (`snakemake --lint`),
* `Snakefile`s are formatted with [snakefmt](https://github.com/snakemake/snakefmt),
* Python wrapper scripts are formatted with [black](https://black.readthedocs.io).
* Conda environments use a minimal amount of channels, in recommended ordering. E.g. for bioconda, use (conda-forge, bioconda, nodefaults, as conda-forge should have highest priority and defaults channels are usually not needed because most packages are in conda-forge nowadays).

### CODE REVIEW

**Comment by pdimens** [comment]
@jrderuiter @johanneskoester there is a silent testing error that still lets the things pass:
```
 =============================== warnings summary ===============================
test.py:134
  /home/runner/work/snakemake-wrappers/snakemake-wrappers/test.py:134: SyntaxWarning: invalid escape sequence '\.'
    "for env in `conda env list | grep -P '\.snakemake/conda' | "
```

File: bio/whatshap/haplotag/wrapper.py [hunk 1]
```diff
@@ -0,0 +1,22 @@
+"""Snakemake wrapper for whatshap haplotag."""
+
+__author__ = "Pavel Dimens"
+__copyright__ = "Copyright 2023, Pavel Dimens"
+__email__ = "pdimens@gmail.com"
+__license__ = "MIT"
+
+from os import path
```
Comment (fgvieira) [inline bio/whatshap/haplotag/wrapper.py:8]: This does not seem to be used

File: bio/whatshap/haplotag/wrapper.py [hunk 2]
```diff
@@ -0,0 +1,22 @@
...
+
+shell(
+    "whatshap haplotag"
+    " {extra}"
+    " --output-threads={snakemake.threads}"
+    " -o {snakemake.output}"
+    " --reference {snakemake.input.reference}"
+    " {snakemake.input.vcf}"
+    " {snakemake.input.bam}"
+)
```
Comment (fgvieira) [inline bio/whatshap/haplotag/wrapper.py:22]: Add the log

File: bio/whatshap/haplotag/meta.yaml [hunk 3]
```diff
@@ -0,0 +1,14 @@
+name: "whatshap haplotag"
+description: Phase BAM records by haplotype. For more information about whatshap see `whatshap documentation <https://whatshap.readthedocs.io/en/latest/guide.html#whatshap-haplotag>`_.
+authors:
```
Comment (fgvieira) [inline bio/whatshap/haplotag/meta.yaml:3]:
```suggestion
url: https://github.com/whatshap/whatshap
authors:
```

File: bio/whatshap/haplotag/meta.yaml [hunk 4]
```diff
@@ -0,0 +1,14 @@
...
+input:
+  - vcf: Path to vcf/vcf.gz/bcf file of phased SNPs.
+  - bam: Path to alignments for the sample in BAM format (must have corresponding .bai index in same directory).
+  - reference: Path to FASTA reference used to create vcf file.
+output: Path to output phased BAM file.
+notes: |
+  If using vcf.gz or bcf files as input, they will require corresponding index files
+  the same directories. Likewise, the input BAM file will require a corresponding .bai
+  index file in the same directory. Specify --ignore-linked-read under extra params if
+  you want whatshap to ignore BX tag information.
```
Comment (fgvieira) [inline bio/whatshap/haplotag/meta.yaml:14]:
```suggestion
  * Input `.vcf.gz` or `.bcf` files requires corresponding index files in the same directories.
  * Input BAM file requires corresponding `.bai` index file in the same directory.
  * Specify `--ignore-linked-read` under `extra` params if
  you want whatshap to ignore `BX` tag information.
```

**Comment by fgvieira** [comment]
Remember to add a test.

File: bio/whatshap/haplotag/wrapper.py (hunk 1 above)
Comment (pdimens) [reply bio/whatshap/haplotag/wrapper.py]: Thanks, this import statement was removed

File: bio/whatshap/haplotag/wrapper.py (hunk 2 above)
Comment (pdimens) [reply bio/whatshap/haplotag/wrapper.py:22]: log added

File: bio/whatshap/haplotag/wrapper.py [hunk 5]
```diff
@@ -0,0 +1,22 @@
...
+
+shell(
+    "whatshap haplotag"
+    "{extra} "
+    "--output-threads={snakemake.threads} "
+    "-o {snakemake.output} "
+    "--reference {snakemake.input.reference} "
+    "{snakemake.input.vcf} "
+    "{snakemake.input.bam} "
+    "{snakemake.log}"
```
Comment (fgvieira) [inline bio/whatshap/haplotag/wrapper.py:21]:
```suggestion
    "{log}"
```

File: bio/whatshap/haplotag/test/Snakefile [hunk 6]
```diff
@@ -0,0 +1,14 @@
+rule whatshap_haplotag:
+    input:
+        vcf="haplotag.10x.vcf.gz",
+        bam="haplotag.10x.bam",
```
Comment (fgvieira) [inline bio/whatshap/haplotag/test/Snakefile:4]: If index files are mandatory, add them as `input`:

```suggestion
        vcf="haplotag.10x.vcf.gz",
        tbi="haplotag.10x.vcf.gz.tbi",
        bam="haplotag.10x.bam",
        bai="haplotag.10x.bam.bai",
```

**Comment by pdimens** [comment]
@fgvieira good point, let me make it more generic

**Comment by pdimens** [comment]
@fgvieira sorry, that's my bad. The `x` was uppercase in the filenames and I didn't realize it :facepalm:

File: bio/whatshap/haplotag/meta.yaml [hunk 7]
```diff
@@ -0,0 +1,21 @@
...
+output: Path to output phased BAM file.
+notes: |
+  * Input `.vcf.gz` or `.bcf` files requires corresponding index files in the same directories.
+  * Input BAM/CRAM file requires corresponding `.bai` (or .crai) index file in the same directory.
+  * Extra params:
+    * --ignore-linked-read (ignore BX tag information)
+    * --linked-read-distance-cutoff DISTANCE (distance after which to split molecules with same BX tag)
+    * --ignore-read-groups (assume all reads come from the same sample)
+    * --sample SAMPLE (only phase this sample from the VCF)
+    * --tag-supplementary (also tag supplementary alignments)
```
Comment (fgvieira) [inline bio/whatshap/haplotag/meta.yaml:21]:
```suggestion
params:
  - extra: additional program arguments (e.g. `--ignore-linked-read` to ignore BX tag, `--linked-read-distance-cutoff` to set distance after which to split molecules with same BX tag, `--ignore-read-groups`, `--sample SAMPLE` to only phase this sample, `--tag-supplementary` to also tag supplementary alignments.
```

File: bio/whatshap/haplotag/wrapper.py [hunk 8]
```diff
@@ -0,0 +1,22 @@
...
+
+extra = snakemake.params.get("extra", "")
+log = snakemake.log_fmt_shell(stdout=False, stderr=True)
+
+shell(
+    "whatshap haplotag"
+    "{extra} "
+    "--output-threads={snakemake.threads} "
+    "-o {snakemake.output} "
+    "--reference {snakemake.input.reference} "
```
Comment (fgvieira) [inline bio/whatshap/haplotag/wrapper.py:18]:
```suggestion
    "--reference {snakemake.input.ref} "
```

File: bio/whatshap/haplotag/test/Snakefile [hunk 9]
```diff
@@ -0,0 +1,16 @@
+rule whatshap_haplotag:
+    input:
+        vcf="haplotag.10X.vcf.gz",
+        vcfindex="haplotag.10X.vcf.gz.tbi",
+        aln="haplotag.10X.bam",
+        alnindex="haplotag.10X.bam.bai", 
+        reference = "reference.fasta"
```
Comment (fgvieira) [inline bio/whatshap/haplotag/test/Snakefile:7]:
```suggestion
        ref = "reference.fasta"
```

File: bio/whatshap/haplotag/meta.yaml [hunk 10]
```diff
@@ -0,0 +1,14 @@
...
+description: Phase BAM records by haplotype. For more information about whatshap see `whatshap documentation <https://whatshap.readthedocs.io/en/latest/guide.html#whatshap-haplotag>`_.
+url: https://github.com/whatshap/whatshap
+authors:
+  - Pavel Dimens
+input:
+  - vcf: Path to vcf.gz/bcf file of phased SNPs.
+  - vcfindex: Path to vcf index file (.tbi for vcf.gz, .csi for .bcf).
+  - aln: Path to alignments for the sample in BAM/CRAM format.
+  - alnindex: Path to alignment index file in .bai/.crai format.
+  - reference: Path to FASTA reference used to create vcf file.
```
Comment (fgvieira) [inline bio/whatshap/haplotag/meta.yaml:11]:
```suggestion
  - ref: Path to FASTA reference used to create vcf file.
```

File: bio/whatshap/haplotag/test/Snakefile [hunk 11]
```diff
@@ -0,0 +1,16 @@
+rule whatshap_haplotag:
+    input:
+        vcf="haplotag.10X.vcf.gz",
+        vcfindex="haplotag.10X.vcf.gz.tbi",
+        aln="haplotag.10X.bam",
+        alnindex="haplotag.10X.bam.bai",
```
Comment (fgvieira) [inline bio/whatshap/haplotag/test/Snakefile:6]:
```suggestion
        vcf="haplotag.10X.vcf.gz",
        aln="haplotag.10X.bam",
        "haplotag.10X.vcf.gz.tbi",
        "haplotag.10X.bam.bai", 
```

File: bio/whatshap/haplotag/meta.yaml [hunk 12]
```diff
@@ -0,0 +1,14 @@
+name: "whatshap haplotag"
+description: Phase BAM records by haplotype. For more information about whatshap see `whatshap documentation <https://whatshap.readthedocs.io/en/latest/guide.html#whatshap-haplotag>`_.
+url: https://github.com/whatshap/whatshap
+authors:
+  - Pavel Dimens
+input:
+  - vcf: Path to vcf.gz/bcf file of phased SNPs.
+  - vcfindex: Path to vcf index file (.tbi for vcf.gz, .csi for .bcf).
+  - aln: Path to alignments for the sample in BAM/CRAM format.
+  - alnindex: Path to alignment index file in .bai/.crai format.
```
Comment (fgvieira) [inline bio/whatshap/haplotag/meta.yaml:10]:
```suggestion
  - Path to vcf index file (.tbi for vcf.gz, .csi for .bcf).
  - aln: Path to alignments for the sample in BAM/CRAM format.
  - Path to alignment index file in .bai/.crai format.
```

**Comment by fgvieira** [comment]
Just small small changes, since I realized that the VCF and BAM index files do not need to be named.

**Comment by pdimens** [comment]
testing failed b/c it was missing the fasta fai file. I added it and included it in the metadata and snakefile.

File: bio/whatshap/haplotag/test/Snakefile [hunk 13]
```diff
@@ -0,0 +1,17 @@
+rule whatshap_haplotag:
+    input:
+        vcf="haplotag.10X.vcf.gz",
+        aln="haplotag.10X.bam",
+        "haplotag.10X.vcf.gz.tbi",
+        "haplotag.10X.bam.bai", 
+        ref = "reference.fasta",
+        "reference.fasta.fai"
```
Comment (fgvieira) [inline bio/whatshap/haplotag/test/Snakefile:8]:
```suggestion
        ref = "reference.fasta",
        "haplotag.10X.vcf.gz.tbi",
        "haplotag.10X.bam.bai", 
        "reference.fasta.fai"
```

File: bio/whatshap/haplotag/test/Snakefile [hunk 14]
```diff
@@ -0,0 +1,17 @@
+rule whatshap_haplotag:
+    input:
+        vcf="haplotag.10X.vcf.gz",
+        aln="haplotag.10X.bam",
+        ref = "reference.fasta",
+        "haplotag.10X.vcf.gz.tbi",
+        "haplotag.10X.bam.bai", 
+        "reference.fasta.fai"
```
Comment (fgvieira) [inline bio/whatshap/haplotag/test/Snakefile:8]:
```suggestion
        "haplotag.10X.vcf.gz.tbi",
        "haplotag.10X.bam.bai", 
        "reference.fasta.fai",
        vcf="haplotag.10X.vcf.gz",
        aln="haplotag.10X.bam",
        ref = "reference.fasta",
```

**Comment by pdimens** [comment]
Odd, I thought I borrowed the right files from the Whatshap repo. I'll have to investigate this tomorrow. Thanks for the ongoing assistance

**Comment by pdimens** [comment]
@fgvieira I replaced the test files with the ones from `data/pacbio/` in the whatshap directory. I tested `whatshap haplotag` with those files locally on my system and hopefully it works this time around.

**Comment by fgvieira** [comment]
Great work! Seems to be working now.
Will merge, if you do not have any other changes.
Thanks for your efforts..

**Comment by pdimens** [comment]
Awesome, thank you! And *sorry* for this not being as straightforward as it could have been. Long time listener, first time caller :grin:

**Pull request merged**